group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -420,6 +422,23 @@ public Group group(String groupId, long committedOffset) throws GroupIdNotFoundE
         return group;
     }
 
+    /**
+     * @return The GenericGroup List filtered by statesFilter and committedOffset.
+     */
+    public List<ListGroupsResponseData.ListedGroup> listGroups(List<String> statesFilter, long committedOffset) {
+        Stream<Group> groupStream = groups.values(committedOffset).stream();
+        if (!statesFilter.isEmpty()) {
+            groupStream = groupStream.filter(group -> {
+                if (group instanceof ConsumerGroup) {
+                    return statesFilter.contains(((ConsumerGroup) group).stateAsString(committedOffset));
+                } else {
+                    return statesFilter.contains(group.stateAsString());
+                }
```

Comment:
This is weird. I would rather push `stateAsString(committedOffset)` to the `Group` interface and use it.

Reply:
make sense.